engiopt/vqgan/vqgan.py

```diff
@@ -0,0 +1,1900 @@
+
+    def __init__(self, channels: int):
+        super().__init__()
+        self.in_channels = channels
+
+        self.gn = GroupNorm(channels)
+        self.q = nn.Conv2d(channels, channels, 1, 1, 0)
+        self.k = nn.Conv2d(channels, channels, 1, 1, 0)
+        self.v = nn.Conv2d(channels, channels, 1, 1, 0)
+        self.proj_out = nn.Conv2d(channels, channels, 1, 1, 0)
+
+    def forward(self, x: th.Tensor) -> th.Tensor:
+        h_ = self.gn(x)
+        q = self.q(h_)
+        k = self.k(h_)
+        v = self.v(h_)
+
+        b, c, h, w = q.shape
+
+        q = q.reshape(b, c, h * w)
+        q = q.permute(0, 2, 1)
+        k = k.reshape(b, c, h * w)
+        v = v.reshape(b, c, h * w)
+
+        attn = th.bmm(q, k)
+        attn = attn * (int(c) ** (-0.5))
+        attn = f.softmax(attn, dim=2)
+        attn = attn.permute(0, 2, 1)
+
+        a = th.bmm(v, attn)
+        a = a.reshape(b, c, h, w)
+
+        return x + a
+
+
+class LinearCombo(nn.Module):
+    """Regular fully connected layer combo for the CVQGAN if enabled.
+
+    Parameters:
+        in_features (int): number of input features
+        out_features (int): number of output features
+        alpha (float): negative slope for LeakyReLU
+    """
+
+    def __init__(self, in_features: int, out_features: int, alpha: float = 0.2):
+        super().__init__()
+        self.model = nn.Sequential(nn.Linear(in_features, out_features), nn.LeakyReLU(alpha))
+
+    def forward(self, x: th.Tensor) -> th.Tensor:
+        return self.model(x)
+
+
+class Encoder(nn.Module):
+    """Encoder module for VQGAN Stage 1.
+
+    Consists of a series of convolutional, residual, and attention blocks arranged using the provided arguments.
+    The number of downsample blocks is determined by the length of the encoder channels tuple minus two.
+    For example, if encoder_channels=(128, 128, 128, 128) and the starting resolution is 128, the encoder will downsample the input image twice, from 128x128 to 32x32.
+
+    Parameters:
+        encoder_channels (tuple[int, ...]): tuple of channel sizes for each encoder layer
+        encoder_start_resolution (int): starting resolution for the encoder
+        encoder_attn_resolutions (tuple[int, ...]): tuple of resolutions at which to apply attention in the encoder
+        encoder_num_res_blocks (int): number of residual blocks per encoder layer
+        image_channels (int): number of channels in the input image
+        latent_dim (int): dimensionality of the latent space
+    """
+
+    def __init__(  # noqa: PLR0913
+        self,
+        encoder_channels: tuple[int, ...],
+        encoder_start_resolution: int,
+        encoder_attn_resolutions: tuple[int, ...],
+        encoder_num_res_blocks: int,
+        image_channels: int,
+        latent_dim: int,
+    ):
+        super().__init__()
+        channels = encoder_channels
+        resolution = encoder_start_resolution
+        layers = [nn.Conv2d(image_channels, channels[0], 3, 1, 1)]
```

Comment:
I would use named args for the Conv2d, it is not straightforward to understand this line without them